Create Invitation API [Validation]

Jtang-1 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
[USH-4757
](https://dimagi.atlassian.net/browse/USH-4757)

The main change in this PR is the addition of an API validation file. This will support the upcoming API for creating Invitations and updating Web Users. The remaining changes involve refactoring to consolidate validations used when bulk uploading Web Users or creating Invitations via the AdminInvitesUserForm.

I reused relevant validations already defined for bulk user import (in user_importer.validation) since the same checks are needed. Some validations previously implemented via the clean method in Django forms were consolidated, while built-in form validations were left as-is.

The key distinction between Invitation creation via the API and via bulk user import is how existing Web Users are handled. Bulk user import checks if the Web User exists and updates it if so. In contrast, the API will have a dedicated POST endpoint for creation. This endpoint will require validating whether the Invitation or Web User already exists and returning an error if it does, mirroring the behavior of the AdminInvitesUserForm.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
No feature flag

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
The main changes (addition of validation for Web User creation API) will not be used yet. The changes to the existing validation code are solely refactors and I tested locally that those areas still work as expected.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
There exists extensive testing on bulk user imports and surrounding validation. `test_importer` checks that web user and mobile workers can be successfully uploaded/updated. `test_validators` test that Location, Profile, Custom Data, etc.. validation catches invalid or unauthorized changes. I will add missing Tableau Group and Role validation.

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
no QA.

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

Jtang-1 committed:
create class to consolidate validation of web user invitation fields

Jtang-1 committed:
this validation is already done as part of django form from CustomDataEditor Form

Jtang-1 committed:
refactor for readability

Jtang-1 committed:
adds function to validate profile

Jtang-1 committed:
refactor validation of email to consolidate

Jtang-1 committed:
do same validation as done for bulk user import in preparation of validating Invitation creation via API

Jtang-1 committed:
refactor: move validating tableau role and tableau group  header to check_headers

Jtang-1 committed:
prefactor - consolidate validation to be used for invitation creation / web user edit API

Jtang-1 committed:
refactor: extract required values from spec in higher level function to make it easier to tell what values are needed

Jtang-1 committed:
create validation for location in preparetion for create invitation/update web user API

Jtang-1 committed:
refactor: location_ids is the equivalent return

Jtang-1 committed:
refactor: consolidate validation for location having users

Jtang-1 committed:
refactor: consolidate validating that primary location is one of the assigned locations or that primary location is defined if assigned locations is defined

Jtang-1 committed:
refactor to extract out validation of tableau groups from the expected format from spec

Jtang-1 committed:
add validation of tableau group to Admin Invites User Validator to be used for invite creation API

Jtang-1 committed:
prefactor: extract out the validation of the tableau_role independent of the spec format in preparation for reuse

Jtang-1 committed:
add validation for tableau role

Jtang-1 committed:
add validation for custom data in preparation for web user creation API

Jtang-1 committed:
remove unnecessary caching

Jtang-1 committed:
refactor: DRY function

Jtang-1 committed:
refactor: The initial purpose of the validation class was to be used for web user API. However, there wasn't as much
overlap with validation used for AdminInvitesUserForm. So move validation for Web User Resource its own file and
create a class to organize the validations relevant to AdminInvitesUserForm

Jtang-1 committed:
perform permission and privilege check only if the parameter exists

Jtang-1 committed:
remove location fields from backend if location shouldn't be shown

Jtang-1 committed:
update test and adds test for validating if invitation email is an existing webuser

Jtang-1 committed:
rename since only one error is returned, not multiple

Jtang-1 committed:
separate conditionally allowed headers since allowed_headers is globally defined and the conditionally allowed headers should not be preserved upon additional calls to check_header

Jtang-1 committed:
fix typo

Jtang-1 committed:
update test

couch_user is already included in the request so the test is updated to reflect that.
The couch_user is now used in `bulk_user_upload_api` so the test needs to be updated

Jtang-1 committed:
test TableauRoleValidation and TableauGroupsValidation

Jtang-1 committed:
fix import

Jtang-1 committed:
test: new tests for validation of Web User Resourcec

Jtang-1 committed:
update test so that order doesn't matter for the listed "illegal column headers"

Jtang-1 committed:
update test so it doesn't matter the order of the arguments

AddisonDunn reviewed (commented):
Love what you're doing here with the refactoring. Was only able to do a partial review so far.

jingcheng16 commented on corehq/apps/registration/forms.py:
Can you link which part of `CustomDataEditor` form already did the validation you deleted?

Jtang-1 replied:
See my reply [here](https://github.com/dimagi/commcare-hq/pull/35448#discussion_r1870199282). The validation is done [here](https://github.com/dimagi/commcare-hq/blob/6fcaa9b6130afba415559877821d5074a312406a/corehq/apps/custom_data_fields/edit_entity.py#L225)

jingcheng16 commented on corehq/apps/registration/validation.py:
```diff
@@ -40,3 +51,12 @@ def validate_profile(self, new_profile_name):
         return profile_validator.validate_spec(spec)
+
+    def validate_email(self, email, is_post):
+        if is_post:
```
Why `is_post` flag is required here? What can happen if we don't have this parameter?

Jtang-1 replied:
The `is_post` is needed in preparation for it being used for the Create Invitation / Update Web User API.  When we're creating a new invitation, we need to check that the email isn’t already associated with an existing user or invitation. But if we’re updating it, we already know the user or invitation exists, so running the same check would always fail.

jingcheng16 commented on corehq/apps/users/validation.py:
```diff
@@ -59,3 +61,16 @@ def validate_profile_required(profile_name, domain):
+
+
+def validate_assigned_locations_has_users(domain, assigned_location_ids):
+    error_message_location_not_has_users = _("These locations cannot have users assigned because of their "
```
Shouldn't we remove `error_message_location_not_has_users` from `LocationValidator` class?

Jtang-1 replied:
yes you're right, thanks
[d3b2d3a](https://github.com/dimagi/commcare-hq/pull/35448/commits/d3b2d3a38f9574d0bf65625d547c341cd6b4b88f)

jingcheng16 reviewed (commented):
Only did partial review. Left a few comments. Nothing blocking.

Jtang-1 committed:
delete unused error message. It is now handled in validate_assigned_locations_has_users

Jtang-1 commented on corehq/apps/registration/forms.py:
```diff
@@ -631,10 +619,17 @@ def clean(self):
 
             if prefixed_profile_key in custom_user_data:
                 profile_id = custom_user_data.pop(prefixed_profile_key)
-                self._validate_profile(profile_id)
```
I don't think so because the validation is already defined on the profile field that's created in the `CustomDataEditor` form [here](https://github.com/dimagi/commcare-hq/blob/6fcaa9b6130afba415559877821d5074a312406a/corehq/apps/custom_data_fields/edit_entity.py#L225). That same field is inserted into `AdminInvitesUserForm` so the validation will be included.

Jtang-1 committed:
correct patch, path should be where function is used

Jtang-1 committed:
For form validation, clean only occurs if request is post so the "is_post" parameter is redundant.

Also, EmailField already builds in EmailValidator. So the additional validation check done in the
bulk import EmailValidator is redundant when cleaning email from the form

Jtang-1 commented on corehq/apps/api/tests/test_validation.py:
```diff
@@ -0,0 +1,105 @@
+        self.assertEqual(self.validator.validate_parameters(params),
+                         "You do not have permission to edit Tableau Configuration.")
+
+    @patch('corehq.apps.registration.validation.domain_has_privilege', return_value=True)
```
Ya, but the patch path should be where the function is used, so it should all be `corehq.apps.registration.validation.domain_has_privilege`. Corrected here [eff4a17](https://github.com/dimagi/commcare-hq/pull/35448/commits/eff4a17b1c6fce40da85ae684f55fdd5ad80857a)

Jtang-1 commented on corehq/apps/registration/forms.py:
```diff
@@ -589,30 +590,17 @@ def __init__(self, data=None, excluded_emails=None, is_add_user=None,
-            raise forms.ValidationError(_("A user with this email address is deactivated. "))
+
+        from corehq.apps.registration.validation import AdminInvitesUserFormValidator
+        error = AdminInvitesUserFormValidator.validate_email(self.domain, email, self.request.method == 'POST')
```
> If an email is being cleaned, isn't this always a POST?

you're right, it's always a POST. Removed the `is_post` parameter [here](https://github.com/dimagi/commcare-hq/pull/35448/commits/6f0ec77e5a09ff9b76e0f79a70aa54cfe1ca9af7). Also, I realized `django.core.validators.validate_email` is already done as part of `EmailField` definition. So I also removed that from `AdminInvitesUserFormValidator.validate_email`

> WebUserInvitationForm needs a better name cause this Form architecture is confusing everytime I look at it. Maybe WebUserInvitationValidationForm.

I'm not sure what you mean by this? I don't think it makes sense to rename `WebUserInvitationForm` to `WebUserInvitationValidationForm`

Jtang-1 replied:
oh i see, I agree that's confusing [06dd234](https://github.com/dimagi/commcare-hq/pull/35448/commits/06dd23401ef7ff306d49b4bf2a9b4184ef8d51b6)

Jtang-1 commented on corehq/apps/registration/validation.py:
```diff
@@ -0,0 +1,45 @@
+            if web_user and not web_user.is_active:
+                return _("A user with this email address is deactivated. ")
+
+        email_validator = EmailValidator(domain, 'email')
```
I removed EmailValidator as part of [this change](https://github.com/dimagi/commcare-hq/pull/35448/commits/6f0ec77e5a09ff9b76e0f79a70aa54cfe1ca9af7). But I do reuse import validators for WebUserResourceValidator, even though I feel a bit unsure about it for the same reasons you mentioned (since it extends `ImportValidator`). Still, I think the validation logic for bulk uploads and the API will probably grow in similar ways, so sharing the logic felt like it made sense. I’m open to other ideas to approach this

Jtang-1 committed:
fix argument

Jtang-1 committed:
fix test - property was moved out of the class

Jtang-1 commented on corehq/apps/users/tests/test_views.py:
```diff
@@ -541,6 +542,52 @@ def test_invalid_workbook_headers(self):
+        with patch('corehq.apps.users.views.mobile.users.BaseUploadUser.upload_users'):
+            response = self._make_post_request(file)
+            self.assertEqual(response.status_code, 200)
+            self.assertEqual(response.json(), {'success': True})
```
As we talked about on slack, I don't plan to use the bulk upload functionality. I plan to implement it as a normal API to be consistent with how it's done for mobile workers

Jtang-1 committed:
avoid regenerating list of roles on each call to reduce memory

AddisonDunn reviewed (commented):
Most of the way there with this PR. This is some cool refactoring.

Jtang-1 committed:
bug: needs to check user has access to existing locations when all locations are being removed locs_ids_being_assigned is empty/falsy)

Jtang-1 committed:
rename function for clarity

Jtang-1 commented on corehq/apps/user_importer/validation.py:
```diff
@@ -513,30 +547,18 @@ def _validate_uploading_user_access(self, spec):
         # to access.
-        if 'location_code' in spec:
-            locs_being_assigned = self._get_locs_being_assigned(spec)
+        if locs_ids_being_assigned:
```
Thank you for catching that bug, fixed it  [a5ee207](https://github.com/dimagi/commcare-hq/pull/35448/commits/a5ee2070e26869c807406c7c234104833dad4b76)

I think this is equivalent - I do the `location_code' in spec` within `validate_spec` and only do the validations if that's true. For the snippet referenced here, I removed the conditional check so that validation will always run.

Jtang-1 replied:
Oh dang, thanks for catching that. [8fee5fe](https://github.com/dimagi/commcare-hq/pull/35448/commits/8fee5fe9462dd5d79b35c6ac5837d01cb0797eff) should fix it. Although I added some nasty looking variable and function names that i'm open to renaming.

Jtang-1 commented on corehq/apps/user_importer/validation.py:
```diff
@@ -150,12 +151,16 @@ class TableauRoleValidator(ImportValidator):
 
     def __init__(self, domain):
         super().__init__(domain)
-        self.valid_role_options = [e.value for e in TableauUser.Roles]
```
Good point, that's preferable to keep it in `__init__` [11edd76](https://github.com/dimagi/commcare-hq/pull/35448/commits/11edd769a1c338a0241457d7e77deeaad0e8bed0)

Jtang-1 commented on corehq/apps/api/validation.py:
```diff
@@ -0,0 +1,87 @@
+            return error
+
+        location_validator = LocationValidator(self.domain, self.requesting_user, self.location_cache, True)
+        location_codes = list(set(assigned_location_codes + [primary_location_code]))
```
IIRC, for both mobile and web user bulk import, it assumes that the first location listed is the primary location. So there's no validation involves, it just assumes and updates/assigns the primary location as such.

Jtang-1 committed:
remove unneccessary model deletion from test database

Jtang-1 commented on corehq/apps/api/validation.py:
```diff
@@ -0,0 +1,87 @@
+        return custom_data_validator.validate_spec(spec)
+
+    def validate_email(self, email, is_post):
+        if is_post:
```
I was confused by that too, but bulk import works a bit differently. Instead of doing validation upfront, it first tries to find the existing user. A new user is only created if no match is found ([see here](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L728)). Similarly, when [processing an invitation](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L843), it [checks for an existing invitation](https://github.com/dimagi/commcare-hq/blob/6366ebf3ef386ef7ecd15d066ecaac738d171496/corehq/apps/user_importer/importer.py#L333) and only creates a new one if it doesn't already exist.

So, the bulk import approach is to try and update existing users or invitations, creating new ones only if they don't exist, rather than validating upfront. This wouldn't work in the case of the API (when POSTing) or the invitation form, though, because those expect to create a new invitation upfront. It wouldn't make sense to fall back to updating an existing invitation if one already exists.

Jtang-1 commented on corehq/apps/api/tests/test_validation.py:
```diff
@@ -0,0 +1,105 @@
+    def test_validate_parameters_without_tableau_edit_permission(self, mock_has_permission):
+        params = {"email": "test@example.com", "role": "Admin", "tableau_role": "Viewer"}
+        self.assertEqual(self.validator.validate_parameters(params),
+                         "You do not have permission to edit Tableau Configuration.")
```
We do, although I refactored it slightly in this PR [here](https://github.com/dimagi/commcare-hq/pull/35448/commits/3d99f3552673ce484492e77188ad6878504e3f4c)

Jtang-1 committed:
clarify naming to distinguish from the form to invite web users

Jtang-1 committed:
correct bug where regardless if locations are being changed,
validation needs to be done that uploading user can access
the current locations of the editable user or invitation.

Jtang-1 committed:
test: tests that invitiation access is checked even if no locations are changed

AddisonDunn reviewed (approved):
Nicely done. Some excellent refactoring done here.

> The changes to the existing validation code are solely refactors and I tested locally that those areas still work as expected

Nice. Given the size of this PR, I reccomend re-testing areas that have had recent commits if you haven't already, and testing the forms that were changed, too (if you haven't already).

Jtang-1 commented:
Thanks for the review @AddisonDunn! I sanity tested admin invite and bulk upload validation surrounding locations and didn't run into any issues.
